### PR TITLE
Setup shaders that render as well

### PR DESCRIPTION
**Pull request by matthijsgroen**
This is still a work in progress!

It is not yet finished. I create this pull request for now to get some early feedback :smile: 

TODO:
- [x] Test this with a built version
- [x] Create setup for images
- [x] Create setup for color
- [x] override default shader for sprites as demo
- [x] Refactor
- [x] Document the whole thing

### CODE REVIEW

**Comment by matthijsgroen** [comment]
It works really great so far.

I can now override the default shader used for each component, and thus override the shader used for sprites.

https://github.com/matthijsgroen/game-play/blob/master/app/scripts/components/shaders/SpriteWithEffects.coffee

overrides the sprite color with same desaturation effect as images. I can now have enemies fly to the background and foreground in a nice way

**Comment by matthijsgroen** [comment]
It's a rough draft for #1001 

There is still a lot to refactor, which I will do later this week.

**Comment by kevinsimper** [comment]
I have not played around with shaders, do you have a good link to read about it? :smile:

**Comment by matthijsgroen** [comment]
@kevinsimper: I've been exploring the code here: https://github.com/greggman/webgl-fundamentals/blob/master/webgl/webgl-2d-image-processing.html

Which is part of this website which I've been reading: http://webglfundamentals.org/

**Comment by matthijsgroen** [comment]
Before I start documenting code, I think it would be better if @starwed or @mucaho did some sort of code review first?

**Comment by matthijsgroen** [comment]
Anybody out there?

**Comment by starwed** [comment]
This looks basically good to me -- it seems quite reasonable to group the various pieces needed for using a shader together in a single object.

I also don't see any issues that this would cause with the (somewhat stalled) work I was doing on adding multiple layers.

**Comment by matthijsgroen** [comment]
@starwed I added documentation. I think it is ready for merging! (Should the whole thing be squashed into a single commit now?)

**Comment by starwed** [comment]
Looks good to me, I'll run a couple of local tests, and if it renders ok go ahead and merge.

**Comment by matthijsgroen** [comment]
:+1:

File: src/graphics/color.js [hunk 1]
```diff
@@ -135,16 +135,25 @@ Crafty.extend({
...
+    fs.readFileSync(__dirname + '/shaders/color.frag', 'utf8'),
+    [
+        { name: "aPosition",    width: 2 },
+        { name: "aOrientation", width: 3 },
+        { name: "aLayer",       width: 2 },
+        { name: "aColor",       width: 4 }
+    ],
+    function(e, _entity) {
+        e.program.writeVector("aColor",
+            entity._red/255,
```
Comment (starwed) [inline src/graphics/color.js:150]: Maybe you mean `entity`, not `_entity` here?  (or vice versa)

(Unfortunately we don't have unit tests for WebGL right now.)

**Comment by starwed** [comment]
Other than the entity/_entity typo in "Color" looks like this works ok.

We don't have unit test coverage for webgl, and somewhat surprisingly `jshint` didn't catch this one either.   

(That said, I've been using [Visual Studio Code](https://code.visualstudio.com/) lately, and it does highlight the use of `entity` as an unknown symbol, which is a nice catch.  For working with Crafty I generally prefer it to Sublime Text out of the box, though I wasn't using many ST extensions.)

**Comment by matthijsgroen** [comment]
It s actually a convention from other programming languages to prefix
unused arguments with an underscore.
Op za 6 feb. 2016 om 21:06 schreef Tim Martin notifications@github.com

**Comment by starwed** [comment]
> It s actually a convention from other programming languages to prefix unused arguments with an underscore.

Hmm, I haven't run into that before -- you do attempt to use the argument without the underscore here, though.

**Comment by matthijsgroen** [comment]
Your correct! I'll fix this later today

**Comment by matthijsgroen** [comment]
Updated. I've merged this version into my game for testing, and everything is fine.

**Comment by starwed** [comment]
Looks good, merging.

**Comment by mucaho** [comment]
Good job :+1: 
Custom shader programs, as well as custom vertex attributes allow for good customization.
For lightning and shadows you need multiple render passes and the ability to render to texture though. Has anyone got some ideas how to achieve that smoothly with current implementation?

**Comment by mucaho** [comment]
Another thing to explore in the future is how to define custom uniforms alongside the internally used ones (unless this is already possible?) - they can be inefficiently emulated by copy-pasting it into every entity's vertex attribute though for now.

**Pull request merged**